pkg/sharing/secret_reconciler.go

```diff
@@ -0,0 +1,178 @@
+package sharing
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+	"time"
+
+	"github.com/go-logr/logr"
+	sgv1alpha1 "github.com/vmware-tanzu/carvel-secretgen-controller/pkg/apis/secretgen/v1alpha1"
+	sgclient "github.com/vmware-tanzu/carvel-secretgen-controller/pkg/client/clientset/versioned"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type SecretReconciler struct {
+	sgClient      sgclient.Interface
+	coreClient    kubernetes.Interface
+	secretExports *SecretExports
+	log           logr.Logger
+}
+
+var _ reconcile.Reconciler = &SecretReconciler{}
+
+func NewSecretReconciler(sgClient sgclient.Interface, coreClient kubernetes.Interface,
+	secretExports *SecretExports, log logr.Logger) *SecretReconciler {
+	return &SecretReconciler{sgClient, coreClient, secretExports, log}
+}
+
+func (r *SecretReconciler) AttachWatches(controller controller.Controller) error {
+	err := controller.Watch(&source.Kind{Type: &corev1.Secret{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return fmt.Errorf("Watching secrets: %s", err)
+	}
+
+	return controller.Watch(&source.Kind{Type: &sgv1alpha1.SecretExport{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(r.mapSecretExportToSecret),
+	})
+}
+
+func (r *SecretReconciler) mapSecretExportToSecret(a handler.MapObject) []reconcile.Request {
+	// TODO expensive call on every secret export update (no cached client used, etc)
```

Comment:
obligatory - this is the TODO you already called out

Reply:
will be addressed in future PRs